File: plugins/collector/event/apps/bigdata/hadoop.go
package bigdata

import (
	"collector/cache/process"
	"collector/event/apps"
	"path/filepath"
	"regexp"
	"strings"
)

type Hadoop struct {
	version    string
	reg        *regexp.Regexp
	regDir     *regexp.Regexp
	regVersion *regexp.Regexp
	mainclass  string
}

func (Hadoop) Name() string { return "hadoop" }

func (Hadoop) Type() string { return "bigdata" }

func (h Hadoop) Version() string { return h.version }

func (h *Hadoop) Match(p *process.Process) bool {
	if p.Name != "java" {
		return false
	}
	if h.reg == nil {
		h.reg = regexp.MustCompile(` org\.apache\.hadoop\.[a-zA-Z\.]+\.DataNode`)
	}
	if s := h.reg.FindString(p.Argv); s == "" {
		return false
	} else {
		h.mainclass = s
	}
	return true
}

func (h *Hadoop) Run(p *process.Process) (m map[string]string, err error) {
	m = make(map[string]string)
	if h.regDir == nil {
		h.regDir = regexp.MustCompile(`hadoop.home.dir=[a-zA-Z0-9\/-_]+`)
	}
	if h.regVersion == nil {
		h.regVersion = regexp.MustCompile(`Hadoop (\d+\.)+\d`)
	}

	var rootPath = "/usr/local/hadoop"
	if s := h.regDir.FindString(p.Argv); s != "" {
		rootPath = strings.TrimPrefix(s, "hadoop.home.dir=")
	}
	if result, err := apps.ExecuteWithName(p, filepath.Join(rootPath, "bin/hadoop"), "version"); err == nil {
		if s := h.regVersion.FindString(result); s != "" {
			h.version = strings.TrimPrefix(s, "Hadoop ")
		}
	}
	m["main_class"] = h.mainclass
	return
}

func init() { apps.Regist(&Hadoop{}) }

File: plugins/collector/src/event/apps/bigdata/hadoop.rs
use crate::event::apps::{execute_with_name, find_version, AppProc, IApp};
use std::collections::HashMap;

pub struct Hadoop {
    version: String,
}

impl Hadoop {
    pub fn new() -> Self { Hadoop { version: String::new() } }
}

impl IApp for Hadoop {
    fn name(&self)     -> &'static str { "hadoop" }
    fn app_type(&self) -> &'static str { "bigdata" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool {
        p.name == "java"
            && p.argv.contains("org.apache.hadoop.")
            && p.argv.contains("DataNode")
    }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        let root = p.argv.split_whitespace()
            .find(|s| s.starts_with("hadoop.home.dir="))
            .map(|s| s.trim_start_matches("hadoop.home.dir="))
            .unwrap_or("/usr/local/hadoop");
        let bin = format!("{}/bin/hadoop", root);
        if let Ok(out) = execute_with_name(p, &bin, &["version"]) {
            // "Hadoop 3.3.4"
            for line in out.lines() {
                if line.starts_with("Hadoop ") {
                    self.version = line.trim_start_matches("Hadoop ").trim().to_owned();
                    break;
                }
            }
            if self.version.is_empty() { self.version = find_version(&out); }
        }
        Ok(HashMap::new())
    }
}
